File: Python/2-Medium/00011._盛最多水的容器/00011._盛最多水的容器.py
from typing import List
def max_area(height: List[int]) -> int:
    """
    计算可以盛放最多水的容器的面积。

    Args:
    height: 一个整数列表，表示每个位置的高度。

    Returns:
    int: 最大的水容量。

    示例:
    >>> max_area([1,8,6,2,5,4,8,3,7])
    49
    """
    # 初始化左右指针和最大面积
    left, right, max_area = 0, len(height) - 1, 0

    # 当左指针小于右指针时，执行循环
    while left < right:
        # 计算当前左右指针构成的容器面积
        current_area = min(height[left], height[right]) * (right - left)
        # 更新最大面积
        max_area = max(max_area, current_area)

        # 移动较短边的指针
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area

# 运行示例
example_height = [1,8,6,2,5,4,8,3,7]
print(f"最大水容量: {max_area(example_height)}")

File: C++/2-Medium/00011._盛最多水的容器/00011._盛最多水的容器.cpp
#include <iostream>
#include <vector>
#include <algorithm> // 用于std::max函数

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0; // 左指针初始化为0
        int right = height.size() - 1; // 右指针初始化为数组的最后一个元素的索引
        int max_area = 0; // 最大面积初始化为0
        
        while (left < right) { // 当左指针小于右指针时循环
            // 计算当前左右指针对应的容器的面积
            int current_area = (right - left) * min(height[left], height[right]);
            max_area = max(max_area, current_area); // 更新最大面积
            
            // 移动指向较矮高度的指针
            if (height[left] < height[right]) {
                ++left; // 左边较矮，移动左指针
            } else {
                --right; // 右边较矮，移动右指针
            }
        }
        
        return max_area; // 返回计算出的最大面积
    }
};

int main() {
    Solution solution;
    vector<int> height = {1,8,6,2,5,4,8,3,7}; // 示例输入
    cout << "Maximum water that can be contained: " << solution.maxArea(height) << endl; // 输出结果
    return 0;
}